Implement config parsing

joebonneau opened the pull request:
- Implements parsing of a sesh/sesh.toml config file and creates a pattern for future use

Open discussion/items:

- Do we like the pattern of parsing the config file in the main cli command? This prevents us from re-running this logic over and over again within a single call, but does introduce a bit of prop drilling that clutters things up a bit.
- The scaffolding to implement the "include root dir in session name" feature is there, but is on hold until we determine whether we want to keep that or not

Here's what the config looks like so far (super barebones):
```toml
[session.name]
include_root_dir = true
```

Also, not sure why my commits from the previous PR are showing up here. If there's some better way to be doing this, let me know... I basically just updated my fork from your repo and then added commits to it.

joshmedeski commented on config/config.go:
```diff
@@ -0,0 +1,45 @@
+)
+
+type SessionName struct {
+	IncludeRootDir bool `toml:"include_root_dir"`
```
Let's call this "PrependParentDir" to be more obvious

joebonneau replied:
Good idea!

joshmedeski commented on session/determine.go:
```diff
@@ -2,15 +2,17 @@ package session
 )
 
-func Determine(choice string) Session {
+func Determine(choice string, config *config.Config) Session {
```
If there's a default config should this be required and not optional?

joebonneau replied:
I'm not sure I follow - Go doesn't really have optional parameters. What you're seeing here is that the Determine function accepts a pointer to a config.Config object to avoid creating copies of the object every time it's drilled down.

joebonneau replied:
Reading your comment again, and I think I just misunderstood your question. I think the reason to always pass the config struct through is because the `ParseConfigFile` function will return a struct with all of the default values, and I imagine that will be useful for conditional logic throughout the various commands.

joshmedeski commented on cmds/choose.go:
```diff
@@ -65,7 +66,8 @@ func Choose() *cli.Command {
 			// TODO: get choice from Session structs array
-			connect.Connect(choice, false, "")
+			config := config.ParseConfigFile()
+			connect.Connect(choice, false, "", &config)
```
Do you like passing all of the config into these functions? I'm not sure how to break up the config file yet but we may end up passing more data than is needed for each function.

joebonneau replied:
I think there are two approaches here: pass exactly what we need, or pass a reference to the config.Config struct. That way, we have all the data that is needed (and then some), but we aren't needlessly creating copies of the config struct. I'm open to either, but I will say that it's maybe a little hard to know at this point what we'll need to pass through for future config values and the parameters could get a bit out of hand if they're not all in a single object that you can pass through.

joshmedeski commented:
I'm checking with Tom about how he uses this parent dir feature. I'm gonna put this PR on hold until I decide we have something work using for a configuration file.

joebonneau commented:
> I'm checking with Tom about how he uses this parent dir feature. I'm gonna put this PR on hold until I decide we have something work using for a configuration file.

Sounds good. Something that I think it could be used for is specifying startup scripts. I'm thinking it could be set up such that you link a path to a startup script wherever you want to store them, and it'll execute that script after creating the session.

All good either way, happy to work on something else entirely in the meantime!

joshmedeski commented:
For #7

joebonneau commented:
@joshmedeski I implemented the startup scripts feature and added some docs to go along with it. For now, users can only run persistent scripts, i.e. the created session won't exit when the script exits. I think this is likely what people will want most of the time anyway, but we can implement the opposite behavior down the road if people are interested.

joshmedeski commented:
Another merge conflict. But for now can we drop the parent session name? I'd like to address #33 and I'll be adding nerd font support soon, which will be need a config option ;)

joshmedeski commented:
Sorry for all the merge conflicts, @joebonneau!

I'll review all of this tonight and hopefully we can get it merged very soon.

joshmedeski commented on connect/connect.go:
```diff
@@ -3,24 +3,22 @@ package connect
 	}
-
-	if err := zoxide.Add(session.Path); err != nil {
+	if err = zoxide.Add(session.Path); err != nil {
```
```suggestion
	if err := zoxide.Add(session.Path); err != nil {
```

We need a `:` here please

joshmedeski reviewed (approved):
Looks good to me 👍
